packages/zoe/src/contractFacet/contractFacet.js

```diff
@@ -0,0 +1,196 @@
+
+import { areRightsConserved } from './rightsConservation';
+import { makeIssuerTable } from '../issuerTable';
+import { assertKeywordName, getKeywords } from '../zoeService/cleanProposal';
+import { evalContractBundle } from './evalContractCode';
+import { makeSeatAdmin } from './seat';
+import { makeExitObj } from './exit';
+
+import '../../exported';
+import '../internal-types';
+
+export function buildRootObject() {
+  /** @type ExecuteContract */
+  const executeContract = async (
+    bundle,
+    zoeService,
+    invitationIssuer,
+    zoeInstanceAdmin,
+    hardenedInstanceRecord,
+  ) => {
+    const instanceRecord = { ...hardenedInstanceRecord };
+    const issuerTable = makeIssuerTable();
+    const getAmountMath = brand => issuerTable.get(brand).amountMath;
+
+    const invitationHandleToHandler = makeWeakStore('invitationHandle');
+    const seatToSeatAdmin = makeWeakStore('seat');
+
+    const issuers = Object.values(instanceRecord.issuerKeywordRecord);
+
+    const getPromiseForIssuerRecords = issuersP =>
+      Promise.all(issuersP.map(issuerTable.getPromiseForIssuerRecord));
+
+    await getPromiseForIssuerRecords(issuers);
+
+    const allSeatStagings = new Set();
+
+    /** @type ContractFacet */
+    const zcf = {
+      reallocate: (/** @type SeatStaging[] */ ...seatStagings) => {
+        // We may want to handle this with static checking instead.
+        // Discussion at: https://github.com/Agoric/agoric-sdk/issues/1017
+        assert(
+          seatStagings.length >= 2,
+          details`reallocating must be done over two or more seats`,
+        );
+
+        seatStagings.forEach(seatStaging =>
+          assert(
+            allSeatStagings.has(seatStaging),
+            details`The seatStaging ${seatStaging} was not recognized`,
+          ),
+        );
+
+        // Ensure that rights are conserved overall. Offer safety was
+        // already checked when an allocation was staged for an individual seat.
+        const flattened = arr => [].concat(...arr);
+        const flattenAllocations = allocations =>
+          flattened(allocations.map(allocation => Object.values(allocation)));
+
+        const previousAllocations = seatStagings.map(seatStaging =>
+          seatStaging.getSeat().getCurrentAllocation(),
+        );
+        const previousAmounts = flattenAllocations(previousAllocations);
+
+        const newAllocations = seatStagings.map(seatStaging =>
+          seatStaging.getStagedAllocation(),
+        );
+        const newAmounts = flattenAllocations(newAllocations);
+
+        assert(
+          areRightsConserved(getAmountMath, previousAmounts, newAmounts),
+          details`Rights are not conserved in the proposed reallocation`,
+        );
+
+        // Commit the staged allocations and inform Zoe of the
+        // newAllocation.
+        seatStagings.forEach(seatStaging =>
+          seatToSeatAdmin.get(seatStaging.getSeat()).commit(seatStaging),
+        );
+      },
```

Comment:
HAZARD NOTE: Would be a safety violation in E or Midori. Is a safety violation for us if messages are sent over unreliable connections, which they're not within a SwingSet.


This commits each seat staging from zcf to zoe individually. Here we call the local `commit` for each new allocation. That local commit does a

```js
      E(zoeSeat).replaceAllocation(currentAllocation);
```
for each one. If the zcf vat dies after only some of the new allocations are committed[1], it can cause Zoe to violate even offer safety, because some old payouts can overdraw the pooled purses for new allocations.

This is also a good example why the three way split is still needed. Because the pooled purses are pooled zoe-wide, a bad crash by one zcf vat can cause offer safety violations in *unrelated* contracts.

[1] It is not a safety violation for us because
   * Messages are only released from completed turns. All these messages are *sent* from the same turn, so either all or none will be emitted from this vat.
   * Within a swingset, or (perhaps?) over ibc, as long as the receiving is alive, every message ever sent to the receiver will eventually be delivered. By contrast, in captp over an unreliable transport, some of these commit messages may be delivered and the rest may not be due to a broken connection.

So even though this program happens to work on our current infrastructure, I do not want to define our computational model to support such delicate assumptions. For E and Midori, neither bullet was true. For Waterken, both were true. I think we should keep the first bullet property (@dtribble may disagree) but I do not think we should guarantee the second in general. Our model should include inter-vat partition. That makes this program incorrect in theory even if it happens to work today in practice.

Reply:
OTOH, if these messages are racing with notification of zcf vat death, could zoe react to a zcf vat death after processing some of these messages but before processing others?

Reply:
Filed at #1391

Reply:
Good point on the analysis. And I agree that we should assume bullet 1 and NOT bullet 2 (eventually). 

So for alpha, I think that the guarantee we have is good enough, since payout liveness in the event of vat failure was not in scope. We just wanted the design to be on the path to that.

I also thought about the racing with the vat failure notification. Because of atomicity of send, I would really like to signal vat failure only after all messages sent from the vat have been delivered. That is consistent for example with there still being messages queued up sent for previous, already-committed turns/cranks.

Reply:
Hi @dtribble thanks. But please add relevant comments to #1391 instead, now that it exists. I'm hitting "Resolve conversation" here.